### PR TITLE
add expand op merge pass and integrate with reshape

### PR DESCRIPTION
**Pull request by Hongqing-work**
### PR Category

CINN

### PR Types

Performance

### Description

Pcard-67164
This PR add a new pass in cinn dialect to fold redundant expand operators. For example, 
`x1 = paddle.expand(x, shape=[4, 16])`
`x2 = paddle.expand(x1, shape=[8, 4, 16]`
can be folded into one expand operator. Besides, if x1 is no more used, the first expand operator can be erased.
Also, this pass is very similar to remove_unchanged_reshape_pass, so reshape and expand folding pass are integrated in this PR.

### CODE REVIEW

File: paddle/cinn/hlir/dialect/operator/transforms/fold_manipulation_ops_pass.cc [hunk 1]
```diff
@@ -85,37 +85,42 @@ class RemoveUnchangedReshapePattern : public pir::OpRewritePattern<OPTYPE> {
...
-  using pir::OpRewritePattern<cinn::dialect::ReshapeOp>::OpRewritePattern;
+  using pir::OpRewritePattern<OPTYPE>::OpRewritePattern;
 
-  bool MatchAndRewrite(cinn::dialect::ReshapeOp op,
+  bool MatchAndRewrite(OPTYPE op,
                        pir::PatternRewriter& rewriter) const override {
-    if (auto pre_shape = op->operand_source(0)
-                             .defining_op()
-                             ->dyn_cast<cinn::dialect::ReshapeOp>()) {
+    if (auto pre_shape = (op->operand_source(0).defining_op())
```
Comment (zyfncg) [inline paddle/cinn/hlir/dialect/operator/transforms/fold_manipulation_ops_pass.cc]: 这里直接叫`pre_op`吧，shape语义上不够准确

File: paddle/cinn/hlir/dialect/operator/transforms/fold_manipulation_ops_pass.cc [hunk 2]
```diff
@@ -85,37 +85,42 @@ class RemoveUnchangedReshapePattern : public pir::OpRewritePattern<OPTYPE> {
...
-    ps.Add<RemoveUnchangedReshapePattern<paddle::dialect::ReshapeOp>>(context);
-    ps.Add<MergeReshapePattern>(context);
+    // remove out_shape equal in_shape ops
+    ps.Add<RemoveUnchangedOpPattern<cinn::dialect::ReshapeOp>>(context);
+    ps.Add<RemoveUnchangedOpPattern<paddle::dialect::ReshapeOp>>(context);
+    ps.Add<RemoveUnchangedOpPattern<cinn::dialect::BroadcastOp>>(context);
+    ps.Add<RemoveUnchangedOpPattern<paddle::dialect::ExpandOp>>(context);
+    // merge redundant ops
+    ps.Add<MergeRedundantOpPattern<cinn::dialect::ReshapeOp>>(context);
+    ps.Add<MergeRedundantOpPattern<cinn::dialect::BroadcastOp>>(context);
```
Comment (zyfncg) [inline paddle/cinn/hlir/dialect/operator/transforms/fold_manipulation_ops_pass.cc:124]: `ExpandOp`是不也需要加上？

File: paddle/cinn/hlir/dialect/operator/transforms/fold_manipulation_ops_pass.cc [hunk 3]
```diff
@@ -85,37 +85,42 @@ class RemoveUnchangedReshapePattern : public pir::OpRewritePattern<OPTYPE> {
...
   }
 };
 
-class RemoveUnchangedReshapePass : public pir::PatternRewritePass {
+class FoldManipulationOpsPass : public pir::PatternRewritePass {
  public:
-  RemoveUnchangedReshapePass()
-      : pir::PatternRewritePass("remove_unchanged_reshape_pass", 1) {}
+  FoldManipulationOpsPass()
+      : pir::PatternRewritePass("fold_manipulationOps_pass", 1) {}
```
Comment (zyfncg) [inline paddle/cinn/hlir/dialect/operator/transforms/fold_manipulation_ops_pass.cc]: `manipulationOps`拆成`manipulation_ops`

File: paddle/cinn/hlir/dialect/operator/transforms/fold_manipulation_ops_pass.cc [hunk 4]
```diff
@@ -126,13 +131,12 @@ class RemoveUnchangedReshapePass : public pir::PatternRewritePass {
...
+  return std::make_unique<FoldManipulationOpsPass>();
 }
-
 }  // namespace ir
 }  // namespace dialect
 }  // namespace cinn
 
-REGISTER_IR_PASS(remove_unchanged_reshape_pass,
-                 ::cinn::dialect::ir::RemoveUnchangedReshapePass);
+REGISTER_IR_PASS(fold_manipulationOps_pass,
```
Comment (zyfncg) [inline paddle/cinn/hlir/dialect/operator/transforms/fold_manipulation_ops_pass.cc]: 同上

Comment (Hongqing-work) [reply paddle/cinn/hlir/dialect/operator/transforms/fold_manipulation_ops_pass.cc]: Done

File: paddle/cinn/hlir/dialect/operator/transforms/fold_manipulation_ops_pass.cc (hunk 3 above)
Comment (Hongqing-work) [reply paddle/cinn/hlir/dialect/operator/transforms/fold_manipulation_ops_pass.cc]: Done

File: paddle/cinn/hlir/dialect/operator/transforms/fold_manipulation_ops_pass.cc (hunk 1 above)
Comment (Hongqing-work) [reply paddle/cinn/hlir/dialect/operator/transforms/fold_manipulation_ops_pass.cc]: Done

File: paddle/cinn/hlir/dialect/operator/transforms/fold_manipulation_ops_pass.cc (hunk 2 above)
Comment (Hongqing-work) [reply paddle/cinn/hlir/dialect/operator/transforms/fold_manipulation_ops_pass.cc:124]: 原本reshape中的merge逻辑中只有cinn::dialect的算子，所以这里针对expand也没添加
理论上将此pass应用到paddle::dialect也没有问题，后续修改已加上了paddle::dialect::ReshapeOp及paddle::dialect::ExpandOp

**Review by zyfncg — APPROVED** [review]

**Review by tc20042008 — APPROVED** [review]

**Pull request merged**